internal/controller/stream_controller.go

```diff
@@ -18,39 +18,371 @@ package controller
-	log.Info("reconcile", "namespace", req.Namespace, "name", req.Name)
-	// TODO(user): your logic here
 
+	if ok := r.ValidNamespace(req.Namespace); !ok {
+		log.Info("Controller restricted to namespace, skipping reconciliation.")
+		return ctrl.Result{}, nil
+	}
+
+	// Fetch stream resource
+	stream := &api.Stream{}
+	if err := r.Get(ctx, req.NamespacedName, stream); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Stream resource not found. Ignoring since object must be deleted.")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("get stream resource '%s': %w", req.NamespacedName.String(), err)
+	}
+
+	// Update ready status to unknown when no status is set
+	if stream.Status.Conditions == nil || len(stream.Status.Conditions) == 0 {
+		log.Info("Setting initial ready condition to unknown.")
+		stream.Status.Conditions = updateReadyCondition(stream.Status.Conditions, v1.ConditionUnknown, "Reconciling", "Starting reconciliation")
+		err := r.Status().Update(ctx, stream)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("set condition unknown: %w", err)
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// Add finalizer
+	if !controllerutil.ContainsFinalizer(stream, streamFinalizer) {
+		log.Info("Adding stream finalizer.")
+		if ok := controllerutil.AddFinalizer(stream, streamFinalizer); !ok {
+			return ctrl.Result{}, errors.New("failed to add finalizer to stream resource")
+		}
+
+		if err := r.Update(ctx, stream); err != nil {
+			return ctrl.Result{}, fmt.Errorf("update stream resource to add finalizer: %w", err)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Check Deletion
+	markedForDeletion := stream.GetDeletionTimestamp() != nil
+	if markedForDeletion {
+		if controllerutil.ContainsFinalizer(stream, streamFinalizer) {
+			err := r.deleteStream(ctx, log, stream)
+			if err != nil {
+				return ctrl.Result{}, fmt.Errorf("delete stream: %w", err)
+			}
+		} else {
+			log.Info("Stream marked for deletion and already finalized. Ignoring.")
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	// Create or update stream
+	if err := r.createOrUpdate(ctx, log, stream); err != nil {
+		return ctrl.Result{}, fmt.Errorf("create or update: %s", err)
+	}
 	return ctrl.Result{}, nil
 }
 
+func (r *StreamReconciler) deleteStream(ctx context.Context, log logr.Logger, stream *api.Stream) error {
+
+	// Set status to not false
+	stream.Status.Conditions = updateReadyCondition(stream.Status.Conditions, v1.ConditionFalse, "Finalizing", "Performing finalizer operations.")
+	if err := r.Status().Update(ctx, stream); err != nil {
+		return fmt.Errorf("update ready condition: %w", err)
+	}
+
+	if !stream.Spec.PreventDelete && !r.ReadOnly() {
+		log.Info("Deleting stream.")
+		err := r.WithJetStreamClient(streamConnOpts(stream.Spec), func(js jetstream.JetStream) error {
+			return js.DeleteStream(ctx, stream.Spec.Name)
+		})
+		if errors.Is(err, jetstream.ErrStreamNotFound) {
+			log.Info("Managed stream was already deleted.")
```

Comment:
Suggestion, don't assume reason for missing stream
```suggestion
			log.Info("Stream does not exist, unable to delete.", "streamName", stream.Spec.Name)
```